pkg/state/state_macvlan_network.go

```diff
@@ -0,0 +1,218 @@
+	return wr
+}
+
+func (s *stateMacvlanNetwork) getManifestObjects(
+	cr *mellanoxv1alpha1.MacvlanNetwork) ([]*unstructured.Unstructured, error) {
+	data := map[string]interface{}{}
+	data["NetworkName"] = cr.Name
+	if cr.Spec.NetworkNamespace == "" {
+		data["NetworkNamespace"] = "default"
+	} else {
+		data["NetworkNamespace"] = cr.Spec.NetworkNamespace
+	}
+
+	data["Master"] = cr.Spec.Master
+	data["Mode"] = cr.Spec.Mode
+	data["Mtu"] = cr.Spec.Mtu
+
+	if cr.Spec.IPAM != "" {
+		data["Ipam"] = "\"ipam\":" + strings.Join(strings.Fields(cr.Spec.IPAM), "")
+	} else {
+		data["Ipam"] = "\"ipam\":{}"
+	}
+
+	// render objects
+	log.V(consts.LogLevelDebug).Info("Rendering objects", "data:", data)
+	objs, err := s.renderer.RenderObjects(&render.TemplatingData{Data: data})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to render objects")
+	}
+	log.V(consts.LogLevelDebug).Info("Rendered", "objects:", objs)
+	return objs, nil
+}
+
+func (s *stateMacvlanNetwork) handleNamespaceChange(cr *mellanoxv1alpha1.MacvlanNetwork,
+	netAttDef *unstructured.Unstructured) error {
+	// Delete NetworkAttachmentDefinition if not in desired namespace
+	lnns, lnnsExists := cr.GetAnnotations()[lastNetworkNamespaceAnnot]
+	netAttDefChangedNamespace := lnnsExists && netAttDef.GetNamespace() != lnns
+	if netAttDefChangedNamespace {
+		err := s.client.Delete(context.TODO(), &netattdefv1.NetworkAttachmentDefinition{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cr.GetName(),
+				Namespace: lnns,
+			},
+		})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *stateMacvlanNetwork) createOrUpdate(cr *mellanoxv1alpha1.MacvlanNetwork,
+	netAttDef *unstructured.Unstructured) error {
+	lnns, lnnsExists := cr.GetAnnotations()[lastNetworkNamespaceAnnot]
+	netAttDefChangedNamespace := lnnsExists && netAttDef.GetNamespace() != lnns
+	if !lnnsExists || netAttDefChangedNamespace {
+		err := s.createObj(netAttDef)
+		if err != nil && !k8serrors.IsAlreadyExists(err) {
+			cr.Status.Reason = "failed to create NetworkAttachmentDefinition"
+			return errors.Wrap(err, "failed to create NetworkAttachmentDefinition")
+		}
+
+		// Update MacvlanNetwork with new netAttDef namespace
+		anno := map[string]string{lastNetworkNamespaceAnnot: netAttDef.GetNamespace()}
+		cr.SetAnnotations(anno)
+		if err := s.client.Update(context.Background(), cr); err != nil {
+			cr.Status.Reason = "failed to update MacvlanNetwork annotations"
+			return errors.Wrap(err, "failed to update MacvlanNetwork annotations")
+		}
+	} else {
+		// CR instances needs ResourceVersion on update
+		currentNetAttDef := netAttDef.DeepCopy()
+		if err := s.getObj(currentNetAttDef); err != nil {
+			cr.Status.Reason = "failed to get NetworkAttachmentDefinition"
+			return errors.Wrap(err, "failed to get NetworkAttachmentDefinition")
+		}
+
+		netAttDef.SetResourceVersion(currentNetAttDef.GetResourceVersion())
```

Comment:
Now that we set ResourceVersion in CreateOrUpdateObjs in `stateSkel` could we leverage this logic here ?

if so please open an issue on it.

Reply:
It will be solved in the rebase and use CreateOrUpdateObjs